### PR TITLE
Split multivariate Student-t distribution sampling

### LINKED ISSUE #12: Implement multivariate Student-t sampler
Student-t distribution has its multivariate generalization (which also for $\nu \gg 2$ is similar to the normal).

1. Sampling can be efficiently implemented as described [here](https://en.wikipedia.org/wiki/Multivariate_t-distribution#Definition).
2. Mutual information can be calculated analytically, as described in
R.B. Arellano-Valle et al., _Shannon Entropy and Mutual Information for Multivariate Skew-Elliptical Distributions_, Scandinavian Journal of Statistics, Vol. 40, No. 1 (March 2013), p. 47.

Moreover, the mentioned article describes MI of several other families of distributions, although sampling from them may be tricky and MI calculation may require numerical integration.

### PR DESCRIPTION
**Pull request by pawel-czyz**
This PR resolves #12 and #19

While testing it I realized that the multinormal sampler didn't have an easy test for MI in the 2D case, so I added it.

### CODE REVIEW

**Comment by pawel-czyz** [comment]
@grfrederic Hmmm, interesting. None of the tests for the MI in the 2D case (which uses fast KSG estimator from SciKit-Learn) doesn't pass that.

Three hypotheses:
1. I introduce a bug to the code. (Either at the MI calculation level or sampling).
2. The formula in the paper has a bug.
3. KSG estimator is not that good for Student-t.

File: src/bmi/samplers/split_student_t.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        """
+        super().__init__(dim_x=dim_x, dim_y=dim_y)
+        # Mutual information of multivariate Student-t contains
+        # the term corresponding to the multivariate normal distribution.
+        # Note that this will also validate all the dimensions
+        # and check whether the dispersion matrix is positive-definite
+        self._multinormal = spl.SplitMultinormal(
+            dim_x=dim_x, dim_y=dim_y, mean=np.zeros_like(mean), covariance=dispersion
+        )
+        self._nu = nu
```
Comment (pawel-czyz) [inline src/bmi/samplers/split_student_t.py:42]: Validate $\nu$.

**Comment by grfrederic** [comment]
with df -> inf we should get a multivariate gaussian, which should pass tests, right? does it? and by should pass i mean that it does pass on our other test.

**Comment by grfrederic** [comment]
since as we talked this is not a priority right now, i think we have two options:
  * merge this branch but don't add automatic tests with these distributions since we know there is some kind of issue
  * leave this PR 
 
personally i lean towards the first option (since it seems the issue could be not numeric/coding), but maybe we should leave some additional warnings for now.

**Comment by grfrederic** [comment]
After testing today with Paweł, it seems that the code works as expected and KSG (and also histogram) MI estimator struggles with computing the correct MI, since the distribution is almost identical to a Gaussian, but has different true MI.

I suggest we make the tests throw warnings not errors and merge. We can talk about why our estimators fail some time later, this might be an interesting case for our paper.

File: src/bmi/samplers/split_student_t.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    Mutual information is based on:
+      R.B. Arellano-Valle, J.E. Contreras-Reyes, M.G. Genton,
+      Shannon Entropy and Mutual Information for Multivariate
+      Skew-Elliptical Distributions,
+      Scandinavian Journal of Statistics, vol. 40, pp. 46-47, 2013
+    """
+
+    def __init__(
+        self, *, dim_x: int, dim_y: int, mean: ArrayLike, dispersion: ArrayLike, df: float
```
Comment (grfrederic) [inline src/bmi/samplers/split_student_t.py:24]: initialize mean to zero if not set

**Review by grfrederic — APPROVED** [review]
as in main thread, change test errors to warnings and it's good to merge ;)

File: src/bmi/samplers/split_student_t.py (hunk 2 above)
Comment (pawel-czyz) [reply src/bmi/samplers/split_student_t.py]: Done :)

File: tests/samplers/test_student_t.py [hunk 3]
```diff
@@ -81,5 +84,30 @@ def test_2d(
     x, y = sampler.sample(n_points=n_samples, rng=10)
     # According to API, y must have shape (n,) rather than (n, 1)
     mi_estimate = mutual_info_regression(x, y.ravel(), random_state=5, n_neighbors=20)[0]
+    true_mi = sampler.mutual_information()
+
+    assert sampler.mutual_information() == pytest.approx(
+        sampler.mi_normal() + sampler.mi_correction()
```
Comment (grfrederic) [inline tests/samplers/test_student_t.py:90]: nice touch!

Comment (pawel-czyz) [reply tests/samplers/test_student_t.py:90]: Thanks! :)

File: src/bmi/samplers/split_student_t.py (hunk 2 above)
Comment (pawel-czyz) [reply src/bmi/samplers/split_student_t.py]: Done :)

**Pull request merged**